### ISSUE TITLE
Relative URIs in dictionaries

### ISSUE THREAD

**Issue by jimsch** [core-wg/coral#44]
Although I have not currently implemented it, I have an assumption that you can have any value as the lookup value.  The question is are CoRIs required to be absolute URIs or are we going to allow them to be relative as well?

Is the following a legal thing to put in as a dictionary value?

399 ==> [5, 0, 6, ".well-known", 6, "core"]

**Comment by ektrah** [comment]
The current idea is that a dictionary can have any value as the lookup value, including relative CoRIs. However, a dictionary entry cannot be referenced from a position where the lookup value is otherwise not allowed.

For example, the draft currently defines:

```cddl
relation-type = text      (text string, draft-ietf-core-coral-01#section-3.1.2)
relation-type /= uint     (dictionary reference, draft-ietf-core-coral-01#section-3.2.1)
```

So you cannot reference (e.g.) a boolean value from a dictionary here, only a text string value. This also applies to relative and absolute CoRIs.

**Comment by ektrah** [comment]
Oh, wait... The draft currently specifies that the lookup value is written in the textual format, i.e., as an IRI reference. Since relative IRI references are not directly convertible to relative CoRIs, you can actually not have a relative CoRI as a dictionary value at the moment...

**Comment by jimsch** [comment]
I was looking at the set of items
 
```
link-target = CoRI / literal
link-target /= #6.TBD6(uint)
```

In this case a relative CoRI would be permitted.

**Comment by jimsch** [comment]
Can you put some text in here?  Do you not believe that the case is real or that it is not permitted?

**Comment by ektrah** [comment]
Sorry, I was under the impression that this issues had been resolved based on your last comment.

**Comment by jimsch** [comment]
Concrete example where a relative URI could be placed in a dictionary.  If you look at the EST document in the ACE working group, there are a number of "pre-defined" well-known paths for EST endpoints, the same is true for the ACE OAuth framework where there are a set of well known paths that are the default values.  This would mean that adding Table 2 of the EST document to the dictionary would allow for these relative URIs to be used via the dictionary rather than using the strings themselves

**Comment by ektrah** [comment]
It seems you want to make all CoRIs in a document relative to some base CoRIs and then check if there's a dictionary value that is equivalent to the resulting relative reference that you could put in place of the URI in the document. As said in https://github.com/core-wg/coral/issues/9#issuecomment-573145754, testing relative references for equivalence in general is difficult. The only reasonable operation defined for a relative CoRI is to resolve it. What you could do instead, is to resolve all dictionary values against the base CoRI and then check if the CoRIs in the document are equivalent to one of the resulting CoRIs:

```diff
 function compress(CoRI, Base, Dictionary) is
-   if exists (K, V) in Dictionary such that V == makeRelative(CoRI, Base) then
+   if exists (K, V) in Dictionary such that resolve(V, Base) == CoRI then
       return K
    else
       return CoRI
```

**Comment by jimsch** [comment]
No, that is the last thing that I want to do.  I am looking at a document along the lines of

AceOauthTokenPost /token 

Both the IRI for AceOauthTokenPost and the CoRI [6, 0, 5, "token"] are in the dictionary so when I encode this in CBOR I get

[1, Dict1, TBD6(Dict2)]

The relative URI in DIct2 is then going to be resolved relative to the document context to get the full URI.  Since 90% of all servers are going to use the same resource path for the token post location, putting it in the dictionary makes sense.  This cannot be encoded as an absolute URI because it is always going to be relative to the base URI of the document.

**Comment by ektrah** [comment]
Yes, as I thought I had clarified above, you can put both `<http://example.com/AceOauthTokenPost>` and `</token>` into the dictionary. What's missing? (Sorry if I keep missing your point :-)

**Comment by jimsch** [comment]
Have you fixed this?

~~~
Oh, wait... The coral-reef draft currently specifies that the lookup value is written in the textual format, i.e., as an IRI reference. Since relative IRI references are not directly convertible to relative CoRIs, you can actually not have a relative CoRI as a dictionary value at the moment...
~~~

**Comment by ektrah** [comment]
No, not yet. I first thought this just the problem of coral-reef, but now I think it’s more general. 

Ideas?

**Comment by jimsch** [comment]
While I agree that it can get very messy, I really want to allow for a value to be a generic CBOR value.  How this gets represented in the IANA registration is a bit problematic.  I don't know if encoded for diagnostic form is better, but I think that one may find that one wants to encode up some "common" value structures to be returned in the event of a private dictionary.  

If this is done, then a CoRI can be directly encoded as a CBOR value.  The only issue comes when one tries to jump back and forth to JSON, but that can be solved by allowing either CBOR or JSON as values with the rule that you do not convert between them.  If you want a "JSON CoRI" (a relative IRI) value to be placed and a normal CoRI value to be placed then both values must be in the dictionary.

**Comment by ektrah** [comment]
Wouldn't it be better to let dictionary values be agnostic of the serialization format?

**Comment by jimsch** [comment]
No I don't think so, the problem is that some values are going to need to be treated differently depending on the serialization format, so unless you are going to try and say how all of those conversions re done then you start having problems.   Additionally, some serializations may be require specific items.  For example a time value may need to use either a tag value of 0 or 1 and that cannot be easily specified for JSON where only one method is permitted.

**Comment by ektrah** [comment]
So, one option would be a dictionary like this:

| Key | CBOR value | JSON value | Textual value |
|-|-|-|-|
| 1234 | `#6.1(1583619415)` | `{ "_time": "2020-03-07T22:16:55Z" }` | `dt'2020-03-07T22:16:55Z'` |
| 4321 | `[0, "coap", 1, "example.com", 5, ".well-known", 5, "core"]` | `{ "_link": "coap://example.com/.well-known/core" }` | `<coap://example.com/.well-known/core>` |

And the other option would be a dictionary like this:

| Key | Abstract value represented in the textual format |
|-|-|
| 1234 | `dt'2020-03-07T22:16:55Z'` |
| 4321 | `<coap://example.com/.well-known/core>` |

...with well-defined conversions from the abstract value into each serialization format.

(Let's ignore for a moment that the JSON and textual serialization formats currently aren't using dictionaries.)

The well-defined conversions from the abstract value into each serialization format are currently defined. There is no ambiguity how `dt'2020-03-07T22:16:55Z'` should be encoded in the CBOR serialization. (It's always tag 1.)

The only limitation with draft-ietf-core-href-02 is that there is conversion of relative URI references to CBOR defined. draft-ietf-core-href-03 will define a (partial) conversion of relative URI references to CBOR, so that limitation goes mostly away.

Given all of this, I still think that storing abstract values (represented in the textual format) in the dictionary is the best solution. Would you agree?

**Comment by jimsch** [comment]
Just where do you believe that this well-defined conversion from abstract values exists?  For the case of a date-time there are three different ways that it can be encoded, and that is before the new date only format and the additional date-time formats that are in a new CBOR draft coming forward. 

There is also not a fixed spec for CBOR diagnostics, it is not specified as a parsable language.  I will ignore the fact that several people including myself have written a parser for it so that is also problematic.  However, I am not sure what happens if you want to put a complex structure here.

Would abstract values be good - sure,  Does it solve the problem?  I am less sure.

**Comment by ektrah** [comment]
> Just where do you believe that this well-defined conversion from abstract values exists? 

From the one-to-one correspondence of the grammars in the CBOR format and the text format:

https://github.com/core-wg/coral/blob/9e96e7a61be810bafc2fa67ae1c500f618b59eb4/binary/grammar.cddl#L19-L21

https://github.com/core-wg/coral/blob/9e96e7a61be810bafc2fa67ae1c500f618b59eb4/textual/syntactic-grammar.abnf#L27-L29

In other words:

| Binary (CDDL) | | Textual (ABNF) |
|-|-|-|
| IRI | :left_right_arrow: | IRI |
| bool | :left_right_arrow: | boolean |
| int | :left_right_arrow: | integer |
| float | :left_right_arrow: | float |
| time | :left_right_arrow: | datetime |
| bytes | :left_right_arrow: | bytes |
| text | :left_right_arrow: | text |
| null | :left_right_arrow: | null |

(where all the above CDDL rules are defined in [RFC 8610 Appendix D](https://tools.ietf.org/html/rfc8610#appendix-D) and the ABNF rules in [I-D.ietf-core-coral-03 Section 4.1.5](https://www.ietf.org/id/draft-ietf-core-coral-03.html#name-literals), with the exception of the `IRI` rules)

> For the case of a date-time there are three different ways that it can be encoded, and that is before the new date only format and the additional date-time formats that are in a new CBOR draft coming forward.

`time`, as referenced above, is defined in RFC 8610 as

```
time = #6.1(number)
```

which is different from

```
tdate = #6.0(tstr)
```

So, with the current draft, there is not really a choice between all the different ways in which you can express date-times in CBOR. (Of course, we could change the draft if we want...)

> There is also not a fixed spec for CBOR diagnostics, it is not specified as a parsable language.

That seems to be an argument against storing CBOR diagnostics in an IANA registry?

> I will ignore the fact that several people including myself have written a parser for it so that is also problematic. However, I am not sure what happens if you want to put a complex structure here.

Currently, the draft specifies that "A dictionary reference MUST NOT refer to a dictionary value that would otherwise not be syntactically allowed in that position."

For example,

https://github.com/core-wg/coral/blob/9e96e7a61be810bafc2fa67ae1c500f618b59eb4/binary/grammar.cddl#L36

So, currently, a dictionary reference in the position of a `link-target` can only reference a `IRI / bool / int / float / time / bytes / text / null`, but no complex structure or anything.

> Would abstract values be good - sure, Does it solve the problem? I am less sure.

I'm not sure I really understand the problem :dizzy_face:

**Comment by jimsch** [comment]
I was thinking that there was something that the grammar does not support.

**Comment by ektrah** [comment]
Yes: the conversions from URI reference to CRI reference and CRI reference to URI reference are only partial.

**Comment by jimsch** [comment]
No, what I was thinking of was replacing an entire set of target values with a dictionary item so that an array of items was done.  It makes sense not to support this.  Still have the relative references to deal with in some way.

**Comment by jimsch** [comment]
So putting something in the dictionary along the lines of MakeRelativeIRI("http://example.org", "http://example.org/foo/bar") would be a text representation of a relative IRI that would have the correct result.
